asf_heat_pump_suitability/pipeline/prepare_features/listed_buildings.py

```diff
@@ -47,18 +69,57 @@ def sjoin_df_epc_with_listed_buildings(
     )
 
     dfs = []
-
     # Group based on the created index
     data_partitioned = epc_df.with_columns(partitions).partition_by("chunk_id")
     logging.info(f"Adding listed buildings to EPC in {len(data_partitioned)} chunks")
     for epc_chunk in tqdm(data_partitioned):
-        epc_gdf = lat_lon.generate_gdf_uprn_coords(df=epc_chunk)[["UPRN", "geometry"]]
-        df = epc_gdf.sjoin(listed_buildings_gdf, how="inner", predicate="intersects")[
-            ["UPRN", "listed_building_grade"]
-        ].drop_duplicates(subset="UPRN")
-
+        df = sjoin_df_epc_listed_buildings(epc_chunk, listed_buildings_gdf)
         dfs.append(df)
 
     df = pl.from_pandas(pd.concat(dfs))
 
+    return df.select(["UPRN", "listed_building"])
+
+
+def sjoin_df_epc_listed_buildings(
+    epc_df: pl.DataFrame, listed_buildings_gdf: gpd.GeoDataFrame, distance: float = 5
+) -> pd.DataFrame:
+    """
+    Spatial join EPC UPRNs with listed buildings using `geopandas.GeoDataFrame.sjoin_nearest` where Point or MultiPoint
+    geometries detected, and `geopandas.GeoDataFrame.sjoin` where Polygons or MultiPolygons detected.
+
+    Args:
+        epc_df (pl.DataFrame): EPC dataset with X and Y coordinates per UPRN
+        listed_buildings_gdf (gpd.GeoDataFrame): listed buildings data
+        distance (float): maximum distance (m) within which to query for nearest geometry where `sjoin_nearest` used.
+                          Must be greater than 0. Default 5.
+
+    Returns:
+        pd.DataFrame: EPC UPRNs in listed buildings
+    """
+    epc_gdf = lat_lon.generate_gdf_uprn_coords(df=epc_df, usecols=["UPRN"])
+    if any(
+        [
+            expr in listed_buildings_gdf.geom_type.unique()
+            for expr in ["Point", "MultiPoint"]
+        ]
+    ):
+        df = epc_gdf.sjoin_nearest(
+            listed_buildings_gdf, how="inner", max_distance=distance
+        )[["UPRN", "listed_building"]].drop_duplicates(subset="UPRN")
+    elif any(
+        [
+            expr in listed_buildings_gdf.geom_type.unique()
+            for expr in ["Polygon", "MultiPolygon"]
+        ]
+    ):
+        df = epc_gdf.sjoin(listed_buildings_gdf, how="inner", predicate="intersects")[
+            ["UPRN", "listed_building"]
+        ].drop_duplicates(subset="UPRN")
```

Comment:
Somehow I ended up responding to this question in the wrong comment reply. Copying my answer here for future ref:

I think (?) it has to be an `sjoin_nearest` because with `sjoin`, these are the predicates for joining:
https://geopandas.org/en/stable/docs/user_guide/mergingdata.html#binary-predicate-joins

intersects
contains
within
touches
crosses
overlaps
So points that aren't exactly touching/ on top of each other will not get joined together. This was the case with the Wales listed buildings `sjoin` that was in the code before. It does join some of the buildings, but you get much more with the `sjoin_nearest` methodology. Does this make sense?